libraries/core_libs/consensus/src/pbft/pbft_manager.cpp

```diff
@@ -1238,26 +1252,68 @@ blk_hash_t PbftManager::proposePbftBlock_() {
   // get DAG block and transaction order
   const auto propose_period = pbft_chain_->getPbftChainSize() + 1;
   auto dag_block_order = dag_mgr_->getDagBlockOrder(dag_block_hash, propose_period);
+
   if (dag_block_order.empty()) {
     LOG(log_er_) << "DAG anchor block hash " << dag_block_hash << " getDagBlockOrder failed in propose";
     assert(false);
   }
-  std::vector<trx_hash_t> non_finalized_transactions;
+
   std::unordered_set<trx_hash_t> trx_hashes_set;
   std::vector<trx_hash_t> trx_hashes;
+  u256 total_weight = 0;
+  uint32_t dag_blocks_included = 0;
   for (auto const &blk_hash : dag_block_order) {
     auto dag_blk = dag_blk_mgr_->getDagBlock(blk_hash);
     if (!dag_blk) {
       LOG(log_er_) << "DAG anchor block hash " << dag_block_hash << " getDagBlock failed in propose for block "
                    << blk_hash;
       assert(false);
     }
-    for (auto const &trx_hash : dag_blk->getTrxs()) {
+    u256 dag_block_weight = 0;
+    const auto &estimations = dag_blk->getEstimations();
+
+    int32_t i = 0;
+    for (const auto &trx_hash : dag_blk->getTrxs()) {
       if (trx_hashes_set.emplace(trx_hash).second) {
         trx_hashes.emplace_back(trx_hash);
+        dag_block_weight += estimations[i];
       }
+      i++;
     }
+    if (total_weight + dag_block_weight > config_.gas_limit) {
+      // we need to form new list of transactions after clipping if block is overweighted
+      trx_hashes.clear();
+      break;
+    }
+    total_weight += dag_block_weight;
+    dag_blocks_included++;
   }
+
+  if (dag_blocks_included != dag_block_order.size()) {
+    auto closest_anchor = findClosestAnchor(ghost, dag_block_order, dag_blocks_included);
+    if (!closest_anchor) {
+      LOG(log_er_) << "Can't find closest anchor after block clipping. Ghost: " << ghost << ". Clipped block_order: "
+                   << vec_blk_t(dag_block_order.begin(), dag_block_order.begin() + dag_blocks_included);
+      assert(false);
+    }
+
+    dag_block_hash = closest_anchor.value();
+    dag_block_order = dag_mgr_->getDagBlockOrder(dag_block_hash, propose_period);
+  }
```

Comment:
I don't think set PBFT gas limit will work. 

There set PBFT "gas_limit": "0x3938700" that is 60000000, and DAG block "gas_limit": "0x989680" that is 10000000.
For example, each DAG block is full (gas fee to 10000000), one PBFT block max include 6 DAG blocks.
If in DAG level 10, there are 8 DAG blocks {b1, ..., b8}, b1 is the anchor in level 10 and have been finalized. And in level 11, b9 is the anchor.
DAG:
...     Level 10       Level 11, ...
        b1                 b9
        b2
        b3 
        b4
        b5
        b6
        b7 
        b8
ghost will be {b9}
dag_block_order will be {b2, b3, b4, b5, b6, b7, b8, b9}
dag_blocks_included will be 6
findClosestAnchor will return {}
Then assertion happen.

So the approach cannot handle the case in each level DAG blocks that have more than PBFT gas limitation

Reply:
But we shouldn't have so much dag blocks in one level I believe. We can change the ratio between this limits to potentially include more dag blocks into pbft if you think thats not enough